### PR TITLE
Backup and restore /etc/products.d (bsc#1219004)

### PR DESCRIPTION
**Pull request by rtamalin**
The Distribution Migration System (DMS) implements a non-interactive
fully automated offline migration of systems by booting into a ISO
image that runs through the migration. In some cases, if a migration
from one product stream to another product stream fails, the normal
rollback handling doesn't always restore the products state correctly.
    
To address this we should include the /etc/products.d contents in the
backup that is created and restored.
    
Add changelog entry for upcoming v1.9.0 release stream, including the
bug tracking id.
    
Add unit test for the zypper Backup and Restore routines.
    
Also fixed two bugs found while developing the unit test:
    
  * CreateTarball()'s check for potential tar arguments incorrectly
    checked under the local file system hierarchy, rather than under
    the specified root hierarchy, meaning it could skip backing up
    entries that might exist under the root if they didn't exist
    locally.
    
  * The /var/adm/backup/system-upgrade directory created under the
    specified root hierarchy, and missing intermediary directories,
    were being created with no access permissions, meaning that only
    a privileged superuser could access them. In reality migrations
    are run as root so it didn't impact customer usage, but it did
    cause problems for the unit tests running as a non-privileged
    user.
    
Fixes: #231, #232
    
Co-authored-by: Felix Schnizlein <felixsch@users.noreply.github.com>

### CODE REVIEW

File: build/packaging/suseconnect-ng.changes [hunk 1]
```diff
@@ -4,6 +4,8 @@ Thu Apr 25 15:39:00 UTC 2024 - Felix Schnizlein <fschnizlein@suse.com>
 - 1.9.0 (unreleased)
   * Build zypper-migration and zypper-packages-search as standalone
     binaries rather then one single binary
+  * Include /etc/products.d in directories whose content are backed
+    up and restored if a zypper-migration rollback happens.
```
Comment (felixsch) [inline build/packaging/suseconnect-ng.changes:8]:
```suggestion
    up and restored if a zypper-migration rollback happens. (bsc#1219004)
```
So maintenance can auto detect the fix for the bug :)

**Comment by felixsch** [comment]
Can you please provide a test case how to test this change? Should be easy I assume but we have our internal ruling to try to provide at least one test case per pull request :)

File: build/packaging/suseconnect-ng.changes (hunk 1 above)
Comment (rtamalin) [reply build/packaging/suseconnect-ng.changes:8]: Oops yes, will do ;-)

**Comment by rtamalin** [comment]
> Can you please provide a test case how to test this change? Should be easy I assume but we have our internal ruling to try to provide at least one test case per pull request :)

Do you mean a feature test or a unit test? There don't appear to be existing unit or feature tests for backup or restore, though I may not be grokking things in the Ruby testing stuff properly...

**Review by rjschwei — APPROVED** [review]

**Review by felixsch — APPROVED** [review]
lgtm :+1:

**How I reviewed this pull request:**

- Ran a migration from `15.3` to `15.5` and checked that both:
  * `/var/adm/backup/system-upgrade/repos.sh`
  * `/var/adm/backup/system-upgrade/repos.tar.gz` exist
- Checked that repos.sh includes `products.d` directory
- Checked that repos.tar.gz includes `products.d/` directory content
- Let the migration rollback and check that everything as been reset

**As always, if you think I missed something I should test, please let me know!**:rocket:

**Pull request merged**